modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/planner/hints/JoinOrderHintsPlannerTest.java

```diff
@@ -0,0 +1,248 @@
+
+        assertPlan(sql, schema, hasChildThat(isInstanceOf(Join.class)
+            .and(j -> j.getJoinType() != JoinRelType.valueOf(joinType))).negate());
+    }
+
+    /**
+     * Tests the commuting of join inputs is disabled by  by {@link HintDefinition#ORDERED_JOINS}.
+     */
+    @Test
+    public void testDisabledCommutingOfJoinInputs() throws Exception {
+        String disabledRules = String.format("DISABLE_RULE('%s', '%s', '%s', '%s')", "MergeJoinConverter",
+            "CorrelatedNestedLoopJoin", JoinPushThroughJoinRule.LEFT.toString(),
+            JoinPushThroughJoinRule.RIGHT.toString());
+
+        String sql = String.format("select /*+ %s, %s */ t3.* from TBL1 t1 JOIN TBL3 t3 on t1.v1=t3.v3 JOIN TBL2 t2 on " +
+            "t2.v2=t1.v1", HintDefinition.ORDERED_JOINS.name(), disabledRules);
+
+        // Tests the plan has no commuted join inputs.
+        assertPlan(sql, schema, nodeOrAnyChild(isInstanceOf(Join.class)
+            .and(input(0, isTableScan("TBL3")))
+            .and(input(1, isTableScan("TBL1")))).negate());
+    }
+
+    /**
+     * Tests join plan building duration. Without enabled forced order, takes too long.
+     */
+    @Test
+    public void testJoinPlanBuildingDuration() throws Exception {
+        // Just a 3-tables join.
+        String sql = "SELECT /*+ " + HintDefinition.ORDERED_JOINS + " */ T1.V1, T2.V1, T2.V2, T3.V1, T3.V2, T3.V3 " +
+            "FROM TBL1 T1 JOIN TBL2 T2 ON T1.V3=T2.V1 JOIN TBL3 T3 ON T2.V3=T3.V1 AND T2.V2=T3.V2";
+
+        long time = 0;
+
+        // Heat a bit and measure only the last run.
+        for (int i = 0; i < 6; ++i) {
+            time = System.nanoTime();
+
+            physicalPlan(sql, schema);
+
+            time = U.nanosToMillis(System.nanoTime() - time);
+
+            log.info("Plan building took " + time + "ms.");
+        }
+
+        assertTrue("Plan building took too long: " + time + "ms.", time < 3000L);
+    }
+
+    /** */
+    @Test
+    public void testDisabledCommutingOfJoinInputsInSubquery() throws Exception {
+        String sqlTpl = "SELECT %s t1.v1, t3.v2 from TBL1 t1 JOIN TBL3 t3 on t1.v3=t3.v3 where t1.v2 in " +
+            "(SELECT %s t2.v2 from TBL2 t2 JOIN TBL3 t3 on t2.v1=t3.v1)";
+
+        // Ensure that the sub-join has inputs order matching the sub-query: 'TBL2->TBL3'.
+        assertPlan(String.format(sqlTpl, "/*+ " + HintDefinition.ORDERED_JOINS + " */", ""), schema,
+            nodeOrAnyChild(isInstanceOf(Join.class).and(input(0, nodeOrAnyChild(isTableScan("TBL1"))))
+                .and(input(1, nodeOrAnyChild(isInstanceOf(Join.class)
+                    .and(input(0, nodeOrAnyChild(isTableScan("TBL2"))))
+                    .and(input(1, nodeOrAnyChild(isTableScan("TBl3")))))))));
+
+        assertPlan(String.format(sqlTpl, "", "/*+ " + HintDefinition.ORDERED_JOINS + " */"), schema,
+            nodeOrAnyChild(isInstanceOf(Join.class).and(input(0, nodeOrAnyChild(isTableScan("TBL1"))))
```

Comment:
Move `.and(input(0, nodeOrAnyChild(isTableScan("TBL1"))))` to the next line to improve readability.